### ISSUE TITLE
[Bug] When consuming messages using messageListener, the ConsumerStats#getMsgNumInReceiverQueue method always returns 0.

### ISSUE THREAD

**Issue by crossoverJie** [apache/pulsar#20076]
### Search before asking

- [X] I searched in the [issues](https://github.com/apache/pulsar/issues) and found nothing similar.

### Version

all version

### Minimal reproduce step

```java
                .messageListener(new MessageListener<byte[]>() {
                    @SneakyThrows
                    @Override
                    public void received(Consumer<byte[]> consumer, Message<byte[]> msg) {
                        TimeUnit.SECONDS.sleep(100);
                        consumer.acknowledge(msg);
                    }
                })

Integer msgNumInReceiverQueue = consumer.getStats().getMsgNumInReceiverQueue();
```
When consuming messages using messageListener, the `msgNumInReceiverQueue` is always 0.

### What did you expect to see?

`consumer.getStats().getMsgNumInReceiverQueue()` returns the actual number of messages that are blocking.

### What did you see instead?

the `msgNumInReceiverQueue` is always 0.

### Anything else?

As all messages are written to an unbounded queue in `externalPinnedExecutor`, should we consider adding a counter here to record the number of unconsumed messages?

### Are you willing to submit a PR?

- [X] I'm willing to submit a PR!

**Comment by tisonkun** [comment]
> `msgNumInReceiverQueue ` returns the actual number of messages that are blocking

I don't think so. When a message arrives to message listener, it's not in the queue. It's users responsibility to track the stats.

**Comment by crossoverJie** [comment]
> > `msgNumInReceiverQueue ` returns the actual number of messages that are blocking
> 
> I don't think so. When a message arrives to message listener, it's not in the queue. It's users responsibility to track the stats.

The user is unable to track these messages as they are still stored in the queue of the `externalPinnedExecutor` thread pool, and the user has no way of knowing how many messages are yet to be consumed.

This means moving the data from the `incomingMessages` queue to the `externalPinnedExecutor` queue.

**Comment by Technoboy-** [comment]
> > > `msgNumInReceiverQueue ` returns the actual number of messages that are blocking
> > 
> > 
> > I don't think so. When a message arrives to message listener, it's not in the queue. It's users responsibility to track the stats.
> 
> The user is unable to track these messages as they are still stored in the queue of the `externalPinnedExecutor` thread pool, and the user has no way of knowing how many messages are yet to be consumed.
> 
> This means moving the data from the `incomingMessages` queue to the `externalPinnedExecutor` queue.

yes, right. Users need to monitor the `externalPinnedExecutor`

**Comment by tisonkun** [comment]
@crossoverJie @Technoboy- Thanks for your clarification! Make sense to me.

Looking forward to your patch 😃
